File: lua/problem_64.lua
local M = {}

function M.knight_tours(n)
    local board = {}
    for i = 1, n do board[i] = {}; for j = 1, n do board[i][j] = false end end
    local moves = {{2,1},{2,-1},{-2,1},{-2,-1},{1,2},{1,-2},{-1,2},{-1,-2}}
    local count = 0
    local function dfs(r, c, visited)
        if visited == n * n then count = count + 1; return end
        for _, m in ipairs(moves) do
            local nr, nc = r + m[1], c + m[2]
            if nr >= 1 and nr <= n and nc >= 1 and nc <= n and not board[nr][nc] then
                board[nr][nc] = true
                dfs(nr, nc, visited + 1)
                board[nr][nc] = false
            end
        end
    end
    for r = 1, n do for c = 1, n do
        board[r][c] = true; dfs(r, c, 1); board[r][c] = false
    end end
    return count
end

function M.problem64()
    print("Problem 64:")
    assert(M.knight_tours(1) == 1)
    assert(M.knight_tours(2) == 0)
    assert(M.knight_tours(3) == 0)
    print("All Tests Passed")
end

return M

File: python/problem_64.py
def knight_tours(n: int) -> int:
    board = [[False] * n for _ in range(n)]
    moves = [(2,1),(2,-1),(-2,1),(-2,-1),(1,2),(1,-2),(-1,2),(-1,-2)]
    count = [0]

    def dfs(r: int, c: int, visited: int):
        if visited == n * n:
            count[0] += 1
            return
        for dr, dc in moves:
            nr, nc = r + dr, c + dc
            if 0 <= nr < n and 0 <= nc < n and not board[nr][nc]:
                board[nr][nc] = True
                dfs(nr, nc, visited + 1)
                board[nr][nc] = False

    for r in range(n):
        for c in range(n):
            board[r][c] = True
            dfs(r, c, 1)
            board[r][c] = False
    return count[0]


def main():
    print("Problem 64:")
    assert knight_tours(1) == 1
    assert knight_tours(2) == 0
    assert knight_tours(3) == 0
    print("All Tests Passed")
